File: app/src/main/java/net/veldor/flibustaloader/workers/DownloadBooksWorker.java
package net.veldor.flibustaloader.workers;

import android.app.Notification;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.work.ForegroundInfo;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.google.common.util.concurrent.ListenableFuture;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.database.AppDatabase;
import net.veldor.flibustaloader.database.dao.BooksDownloadScheduleDao;
import net.veldor.flibustaloader.database.dao.DownloadedBooksDao;
import net.veldor.flibustaloader.database.entity.BooksDownloadSchedule;
import net.veldor.flibustaloader.database.entity.DownloadedBooks;
import net.veldor.flibustaloader.ecxeptions.BookNotFoundException;
import net.veldor.flibustaloader.ecxeptions.ConnectionLostException;
import net.veldor.flibustaloader.ecxeptions.TorNotLoadedException;
import net.veldor.flibustaloader.http.ExternalVpnVewClient;
import net.veldor.flibustaloader.http.TorWebClient;
import net.veldor.flibustaloader.notificatons.Notificator;
import net.veldor.flibustaloader.ui.BaseActivity;
import net.veldor.flibustaloader.utils.FilesHandler;
import net.veldor.flibustaloader.utils.MyPreferences;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import static net.veldor.flibustaloader.notificatons.Notificator.DOWNLOAD_PROGRESS_NOTIFICATION;
import static net.veldor.flibustaloader.view_models.MainViewModel.MULTIPLY_DOWNLOAD;

public class DownloadBooksWorker extends Worker {
    private final Notificator mNotificator;
    public static boolean downloadInProgress = false;
    public DownloadBooksWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
        mNotificator = App.getInstance().getNotificator();
    }

    public static void dropDownloadsQueue() {
        // удалю из базы данных всю очередь скачивания
        AppDatabase db = App.getInstance().mDatabase;
        BooksDownloadScheduleDao dao = db.booksDownloadScheduleDao();
        List<BooksDownloadSchedule> schedule = dao.getAllBooks();
        if (schedule != null && schedule.size() > 0) {
            for (BooksDownloadSchedule b : schedule) {
                dao.delete(b);
            }
        }
        // уведомлю, что размер списка закачек изменился
        BaseActivity.sLiveDownloadScheduleCount.postValue(true);
    }

    public static void removeFromQueue(BooksDownloadSchedule scheduleItem) {
        AppDatabase db = App.getInstance().mDatabase;
        BooksDownloadScheduleDao dao = db.booksDownloadScheduleDao();
        dao.delete(scheduleItem);
        // уведомлю, что размер списка закачек изменился
        BaseActivity.sLiveDownloadScheduleCount.postValue(true);
    }

    public static boolean noActiveDownloadProcess() {
        // проверю наличие активных процессов скачивания
        ListenableFuture<List<WorkInfo>> info = WorkManager.getInstance(App.getInstance()).getWorkInfosForUniqueWork(MULTIPLY_DOWNLOAD);
        try {
            List<WorkInfo> results = info.get();
            if (results == null || results.size() == 0) {
                return true;
            }
            for (WorkInfo v : results) {
                WorkInfo.State status = v.getState();
                if (status.equals(WorkInfo.State.ENQUEUED) || status.equals(WorkInfo.State.RUNNING)) {
                    return false;
                }
            }
        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return true;
    }

    public static LiveData<List<WorkInfo>> getDownloadProgress() {
        return WorkManager.getInstance(App.getInstance()).getWorkInfosForUniqueWorkLiveData(MULTIPLY_DOWNLOAD);
    }

    public static void skipFirstBook() {
        AppDatabase db = App.getInstance().mDatabase;
        BooksDownloadScheduleDao dao = db.booksDownloadScheduleDao();
        dao.delete(dao.getFirstQueuedBook());
        // уведомлю, что размер списка закачек изменился
        BaseActivity.sLiveDownloadScheduleCount.postValue(true);
    }

    @NonNull
    @Override
    public Result doWork() {
        downloadInProgress = true;
        long downloadStartTime = System.currentTimeMillis();
        if (App.getInstance().useMirror) {
            // оповещу о невозможности скачивания книг с альтернативного зеркала
            mNotificator.notifyDownloadFromMirror();
        }
        ArrayList<BooksDownloadSchedule> downloadErrors = new ArrayList<>();
        // проверю, есть ли в очереди скачивания книги
        AppDatabase db = App.getInstance().mDatabase;
        BooksDownloadScheduleDao dao = db.booksDownloadScheduleDao();
        DownloadedBooksDao downloadBooksDao = db.downloadedBooksDao();
        Boolean reDownload = App.getInstance().isReDownload();
        try {
            // получу количество книг на начало скачивания
            int mBooksCount = dao.getQueueSize();
            int bookDownloadsWithErrors = 0;
            if (mBooksCount > 0) {
                // помечу рабочего важным
                ForegroundInfo info = createForegroundInfo();
                setForegroundAsync(info);
                // создам уведомление о скачивании
                mNotificator.mDownloadScheduleBuilder.setProgress(mBooksCount, 0, true);
                mNotificator.mNotificationManager.notify(DOWNLOAD_PROGRESS_NOTIFICATION, mNotificator.mDownloadScheduleBuilder.build());
                BooksDownloadSchedule queuedElement;
                // начну скачивание
                // периодически удостовериваюсь, что работа не отменена
                if (isStopped()) {
                    // немедленно прекращаю работу
                    mNotificator.cancelBookLoadNotification();
                    downloadInProgress = false;
                    return Result.success();
                }
                int downloadCounter = 1;
                // пока есть книги в очереди скачивания и работа не остановлена
                while ((queuedElement = dao.getFirstQueuedBook()) != null && !isStopped()) {
                    queuedElement.name = queuedElement.name.replaceAll("\\p{C}", "");
                    mNotificator.updateDownloadProgress(dao.getQueueSize() + downloadCounter - 1, downloadCounter, downloadStartTime);
                    // проверю, не загружалась ли уже книга, если загружалась и запрещена повторная загрузка- пропущу её
                    if (!reDownload && downloadBooksDao.getBookById(queuedElement.bookId) != null) {
                        dao.delete(queuedElement);
                        // уведомлю, что размер списка закачек изменился
                        BaseActivity.sLiveDownloadScheduleCount.postValue(true);
                        continue;
                    }
                    // загружу книгу
                    try {
                        downloadBook(queuedElement);
                        if (!isStopped()) {
                            if (queuedElement.loaded) {
                                // отмечу книгу как скачанную
                                DownloadedBooks downloadedBook = new DownloadedBooks();
                                downloadedBook.bookId = queuedElement.bookId;
                                downloadBooksDao.insert(downloadedBook);
                                // удалю книгу из очереди скачивания
                                // покажу уведомление о успешной загрузке
                                mNotificator.sendLoadedBookNotification(queuedElement);
                                dao.delete(queuedElement);
                                // уведомлю, что размер списка закачек изменился
                                BaseActivity.sLiveDownloadScheduleCount.postValue(true);
                                // оповещу о скачанной книге
                                App.getInstance().mLiveDownloadedBookId.postValue(queuedElement.bookId);
                                // если не клянчил донаты- поклянчу :)
                                if (!MyPreferences.getInstance().askedForDonation()) {
                                    mNotificator.begDonation();
                                    MyPreferences.getInstance().setDonationBegged();
                                }
                            } else {
                                Log.d("surprise", "DownloadBooksWorker doWork 178: book not loaded or loaded with zero size");
                                mNotificator.sendBookNotFoundInCurrentFormatNotification(queuedElement);
                                bookDownloadsWithErrors++;
                                downloadErrors.add(queuedElement);
                                dao.delete(queuedElement);
                            }
                        }
                    } catch (BookNotFoundException e) {
                        Log.d("surprise", "DownloadBooksWorker doWork 173: catch book not found error");
                        // ошибка загрузки книг, выведу сообщение об ошибке
                        mNotificator.sendBookNotFoundInCurrentFormatNotification(queuedElement);
                        bookDownloadsWithErrors++;
                        downloadErrors.add(queuedElement);
                        dao.delete(queuedElement);
                        // уведомлю, что размер списка закачек изменился
                        BaseActivity.sLiveDownloadScheduleCount.postValue(true);
                    } catch (TorNotLoadedException e) {
                        Log.d("surprise", "DownloadBooksWorker doWork 172: catch tor load exception, download work stopped");
                        e.printStackTrace();
                        // при ошибке загрузки TOR остановлю работу
                        if (downloadErrors.size() > 0) {
                            for (BooksDownloadSchedule b :
                                    downloadErrors) {
                                dao.insert(b);
                                downloadErrors.remove(b);
                            }
                        }
                        mNotificator.cancelBookLoadNotification();
                        mNotificator.showTorNotLoadedNotification();
                        downloadInProgress = false;
                        return Result.success();
                    }
                    ++downloadCounter;
                }
                // цикл закончился, проверю, все ли книги загружены
                mBooksCount = dao.getQueueSize();
                if (mBooksCount == 0 && !isStopped()) {
                    // ура, всё загружено, выведу сообщение об успешной загрузке
                    mNotificator.showBooksLoadedNotification(bookDownloadsWithErrors);
                    // Добавлю все книги с ошибками обратно в список загрузки
                    for (BooksDownloadSchedule b :
                            downloadErrors) {
                        dao.insert(b);
                        downloadErrors.remove(b);
                    }
                    // уведомлю, что размер списка закачек изменился
                    BaseActivity.sLiveDownloadScheduleCount.postValue(true);
                }
            }
        } finally {
            if (downloadErrors.size() > 0) {
                for (BooksDownloadSchedule b :
                        downloadErrors) {
                    dao.insert(b);
                    downloadErrors.remove(b);
                }
            }
        }
        mNotificator.cancelBookLoadNotification();
        downloadInProgress = false;
        return Result.success();
    }

    private void downloadBook(BooksDownloadSchedule book) throws BookNotFoundException, TorNotLoadedException {
        if (App.getInstance().isExternalVpn()) {
            Log.d("surprise", "DownloadBooksWorker downloadBook try download trough external vpn");
            ExternalVpnVewClient.downloadBook(book);
        } else {
            // настрою клиент
            try {
                TorWebClient client = new TorWebClient();
                client.downloadBook(book);
            } catch (ConnectionLostException e) {
                mNotificator.showTorNotLoadedNotification();
                e.printStackTrace();
                throw new TorNotLoadedException();
            }
        }
    }


    @NonNull
    private ForegroundInfo createForegroundInfo() {
        // Build a notification
        Notification notification = mNotificator.createMassBookLoadNotification();
        return new ForegroundInfo(DOWNLOAD_PROGRESS_NOTIFICATION, notification);
    }
}

File: app/src/main/java/net/veldor/flibustaloader/workers/DownloadBooksWorker.kt
package net.veldor.flibustaloader.workers

import android.content.Context
import android.util.Log
import androidx.work.ForegroundInfo
import androidx.work.Worker
import androidx.work.WorkerParameters
import net.veldor.flibustaloader.App
import net.veldor.flibustaloader.database.entity.BooksDownloadSchedule
import net.veldor.flibustaloader.database.entity.DownloadedBooks
import net.veldor.flibustaloader.ecxeptions.BookNotFoundException
import net.veldor.flibustaloader.ecxeptions.DownloadsDirNotFoundException
import net.veldor.flibustaloader.ecxeptions.TorNotLoadedException
import net.veldor.flibustaloader.handlers.LoadedBookHandler
import net.veldor.flibustaloader.http.ExternalVpnVewClient
import net.veldor.flibustaloader.http.TorWebClient
import net.veldor.flibustaloader.notificatons.NotificationHandler
import net.veldor.flibustaloader.selections.CurrentBookDownloadProgress
import net.veldor.flibustaloader.selections.TotalBookDownloadProgress
import net.veldor.flibustaloader.ui.BaseActivity
import net.veldor.flibustaloader.utils.Grammar
import net.veldor.flibustaloader.utils.PreferencesHandler
import net.veldor.flibustaloader.utils.RandomString
import net.veldor.flibustaloader.utils.URLHelper
import net.veldor.flibustaloader.view_models.DownloadScheduleViewModel
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStream
import java.util.*

class DownloadBooksWorker(
    context: Context,
    workerParams: WorkerParameters
) :
    Worker(context, workerParams) {

    private val downloadErrors: ArrayList<BooksDownloadSchedule> = arrayListOf()

    override fun doWork(): Result {
        App.instance.liveDownloadState.postValue(DOWNLOAD_IN_PROGRESS)
        val downloadStartTime = System.currentTimeMillis()
        // проверю, есть ли в очереди скачивания книги
        val db = App.instance.mDatabase
        val dao = db.booksDownloadScheduleDao()
        val downloadBooksDao = db.downloadedBooksDao()
        val reDownload = PreferencesHandler.instance.isReDownload
        try {
            // получу количество книг на начало скачивания
            var booksCount = dao.queueSize
            var bookDownloadsWithErrors = 0
            var booksDownloadedYet = 0
            var totalBooksCount: Int
            if (booksCount > 0) {
                // помечу рабочего важным
                val info = createForegroundInfo()
                setForegroundAsync(info)
                var queuedElement: BooksDownloadSchedule?
                // начну скачивание
                // периодически удостовериваюсь, что работа не отменена
                if (isStopped) {
                    // немедленно прекращаю работу
                    NotificationHandler.instance.cancelBookLoadNotification()
                    App.instance.liveDownloadState.postValue(DOWNLOAD_FINISHED)
                    return Result.success()
                }
                var downloadCounter = 1
                // получу оставшееся количество книг для загрузки
                totalBooksCount = booksDownloadedYet + dao.queueSize

                var progress = TotalBookDownloadProgress()
                progress.total = totalBooksCount
                progress.loaded = downloadCounter
                progress.failed = bookDownloadsWithErrors
                DownloadScheduleViewModel.liveFullBookDownloadProgress.postValue(progress)
                NotificationHandler.instance.updateDownloadProgress(
                    totalBooksCount,
                    downloadCounter,
                    downloadStartTime
                )
                while (true) {
                    if (isStopped) {
                        // немедленно прекращаю работу
                        NotificationHandler.instance.cancelBookLoadNotification()
                        App.instance.liveDownloadState.postValue(DOWNLOAD_FINISHED)
                        return Result.success()
                    }
                    // получу первый элемент из очереди
                    queuedElement = dao.firstQueuedBook
                    if (queuedElement == null || isStopped) {
                        break
                    }
                    // получу оставшееся количество книг для загрузки
                    totalBooksCount = booksDownloadedYet + dao.queueSize

                    queuedElement.name = queuedElement.name.replace("\\p{C}".toRegex(), "")

                    progress = TotalBookDownloadProgress()
                    progress.total = totalBooksCount
                    progress.loaded = downloadCounter
                    progress.failed = bookDownloadsWithErrors
                    DownloadScheduleViewModel.liveFullBookDownloadProgress.postValue(progress)
                    NotificationHandler.instance.updateDownloadProgress(
                        totalBooksCount,
                        downloadCounter,
                        downloadStartTime
                    )

                    // проверю, не загружалась ли уже книга, если загружалась и запрещена повторная загрузка- пропущу её
                    if (!reDownload && downloadBooksDao.getBookById(queuedElement.bookId) != null) {
                        dao.delete(queuedElement)
                        // уведомлю, что размер списка закачек изменился
                        BaseActivity.sLiveDownloadScheduleCountChanged.postValue(true)
                        continue
                    }
                    // загружу книгу
                    try {
                        App.instance.liveBookDownloadInProgress.postValue(queuedElement)
                        downloadBook(queuedElement)
                        if (!isStopped) {
                            if (queuedElement.loaded) {
                                // отмечу книгу как скачанную
                                val downloadedBook = DownloadedBooks()
                                downloadedBook.bookId = queuedElement.bookId
                                downloadBooksDao.insert(downloadedBook)
                                // удалю книгу из очереди скачивания
                                dao.delete(queuedElement)
                                // уведомлю, что размер списка закачек изменился
                                BaseActivity.sLiveDownloadScheduleCountChanged.postValue(true)
                                // оповещу о скачанной книге
                                App.instance.liveBookJustLoaded.postValue(queuedElement)
                                App.instance.mLiveDownloadedBookId.postValue(queuedElement.bookId)
                                // если не клянчил донаты- поклянчу :)
                                if (!PreferencesHandler.instance.askedForDonation()) {
                                    NotificationHandler.instance.begDonation()
                                    PreferencesHandler.instance.setDonationBegged()
                                }
                            } else {
                                NotificationHandler.instance.sendBookNotFoundInCurrentFormatNotification(
                                    queuedElement
                                )
                                App.instance.liveBookJustError.postValue(queuedElement)
                                bookDownloadsWithErrors++
                                downloadErrors.add(queuedElement)
                                dao.delete(queuedElement)
                            }
                        }
                    } catch (e: Exception) {
                        Log.d(
                            "surprise",
                            "DownloadBooksWorker doWork 173: catch book not found error"
                        )
                        e.printStackTrace()
                        // ошибка загрузки книг, выведу сообщение об ошибке
                        App.instance.liveBookJustError.postValue(queuedElement)
                        NotificationHandler.instance.sendBookNotFoundInCurrentFormatNotification(
                            queuedElement
                        )
                        bookDownloadsWithErrors++
                        downloadErrors.add(queuedElement)
                        dao.delete(queuedElement)
                        // уведомлю, что размер списка закачек изменился
                        BaseActivity.sLiveDownloadScheduleCountChanged.postValue(true)
                    }
                    ++downloadCounter
                    ++booksDownloadedYet
                }
                // цикл закончился, проверю, все ли книги загружены
                booksCount = dao.queueSize
                if (booksCount == 0 && !isStopped) {
                    // ура, всё загружено, выведу сообщение об успешной загрузке
                    NotificationHandler.instance.showBooksLoadedNotification(bookDownloadsWithErrors)
                }
            }
        } finally {
            if (!downloadErrors.isNullOrEmpty()) {
                downloadErrors.forEach {
                    dao.insert(it)
                }
                BaseActivity.sLiveDownloadScheduleCountChanged.postValue(true)
            }
        }
        NotificationHandler.instance.cancelBookLoadNotification()
        App.instance.liveDownloadState.postValue(DOWNLOAD_FINISHED)
        return Result.success()
    }


    @Throws(
        BookNotFoundException::class,
        TorNotLoadedException::class,
        DownloadsDirNotFoundException::class
    )
    private fun downloadBook(book: BooksDownloadSchedule): Boolean {
        val base = URLHelper.getBaseUrl()
        var link = book.link
        if (base.endsWith("/") && link.startsWith("/")) {
            link = link.substring(1)
        }
        val startTime = System.currentTimeMillis()
        val bookUrl = base + link
        Log.d("surprise", "downloadBook: load $bookUrl")
        // получу response доступным способом
        val response =
            if (PreferencesHandler.instance.isExternalVpn) ExternalVpnVewClient.rawRequest(bookUrl) else TorWebClient().rawRequest(
                bookUrl
            )
        if (response != null) {
            // проверю, что запрос успешен
            if (response.statusLine.statusCode in 200..310) {
                if (book.name.isEmpty()) {
                    // try to get file name with extension
                    val filenameHeader = response.getLastHeader("Content-Disposition")
                    if (filenameHeader != null) {
                        book.name = Grammar.removeExtension(
                            filenameHeader.value.replace(
                                "attachment; filename=",
                                ""
                            )
                        )
                    }
                }
                // ответ сервера получен
                // загружу временный файл
                val content = response.entity.content
                val contentLength = response.entity.contentLength
                if (content != null && response.entity.contentLength > 0) {
                    val tempFile = File.createTempFile(RandomString().nextString(), null)
                    tempFile.deleteOnExit()
                    val out: OutputStream = FileOutputStream(tempFile)
                    var read: Int
                    val buffer = ByteArray(1024)
                    while (content.read(buffer).also { read = it } > 0) {
                        if (isStopped) {
                            NotificationHandler.instance.cancelBookLoadingProgressNotification()
                            // немедленно прекращаю работу
                            NotificationHandler.instance.cancelBookLoadNotification()
                            App.instance.liveDownloadState.postValue(DOWNLOAD_FINISHED)
                            return false
                        }
                        out.write(buffer, 0, read)


                        val progress = CurrentBookDownloadProgress()
                        progress.fullSize = contentLength
                        progress.loadedSize = tempFile.length()
                        DownloadScheduleViewModel.liveCurrentBookDownloadProgress.postValue(progress)
                        if (PreferencesHandler.instance.showDownloadProgress) {
                            NotificationHandler.instance.createBookLoadingProgressNotification(
                                contentLength.toInt(),
                                tempFile.length().toInt(),
                                book.name,
                                startTime
                            )
                        }
                    }
                    out.close()
                    content.close()
                    NotificationHandler.instance.cancelBookLoadingProgressNotification()
                    if (tempFile.length() > 1000) {
                        Log.d(
                            "surprise",
                            "downloadBook: loaded something with content length ${tempFile.length()} and status ${response.statusLine.statusCode}"
                        )
                        // книга загружена, помещу её в нужную папку и правильно назову
                        if (isStopped) {
                            // немедленно прекращаю работу
                            NotificationHandler.instance.cancelBookLoadNotification()
                            App.instance.liveDownloadState.postValue(DOWNLOAD_FINISHED)
                            return false
                        }
                        LoadedBookHandler().saveBook(book, response, tempFile)
                        book.loaded = true
                        return true
                    }
                }
            }
        }
        return false
    }

    private fun createForegroundInfo(): ForegroundInfo {
        // Build a notification
        val notification = NotificationHandler.instance.createMassBookLoadNotification()
        return ForegroundInfo(NotificationHandler.DOWNLOAD_PROGRESS_NOTIFICATION, notification)
    }

    companion object {
        @JvmStatic
        fun dropDownloadsQueue() {
            // удалю из базы данных всю очередь скачивания
            val db = App.instance.mDatabase
            val dao = db.booksDownloadScheduleDao()
            val schedule = dao.allBooks
            if (schedule != null && schedule.isNotEmpty()) {
                for (b in schedule) {
                    dao.delete(b)
                }
            }
            // уведомлю, что размер списка закачек изменился
            BaseActivity.sLiveDownloadScheduleCountChanged.postValue(true)
        }

        @JvmStatic
        fun removeFromQueue(scheduleItem: BooksDownloadSchedule?) {
            val db = App.instance.mDatabase
            val dao = db.booksDownloadScheduleDao()
            dao.delete(scheduleItem)
            // уведомлю, что размер списка закачек изменился
            BaseActivity.sLiveDownloadScheduleCountChanged.postValue(true)
        }

        @JvmStatic
        fun skipFirstBook() {
            val db = App.instance.mDatabase
            val dao = db.booksDownloadScheduleDao()
            dao.delete(dao.firstQueuedBook)
            // уведомлю, что размер списка закачек изменился
            BaseActivity.sLiveDownloadScheduleCountChanged.postValue(true)
        }

        const val DOWNLOAD_IN_PROGRESS = "download in progress"
        const val DOWNLOAD_FINISHED = "download finished"
    }

    override fun onStopped() {
        super.onStopped()
        Log.d("surprise", "onStopped: download interrupted")
    }
}